pkg/node-port-handler/node-port-handler.go

```diff
@@ -0,0 +1,120 @@
+	factory := informers.NewSharedInformerFactory(clientset, 5*time.Minute)
+	informer := factory.Core().V1().Services().Informer()
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			service := obj.(*v1.Service)
+			log.Infof("Added service '%s' of type '%s'", service.GetName(), service.Spec.Type)
+			if service.Spec.Type != v1.ServiceTypeNodePort {
+				log.Debugf("Service '%s' is not of type NodePort. Nothing to do.", service.GetName())
+				return
+			}
+
+			for _, port := range service.Spec.Ports {
+				expose(port.NodePort)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldService := oldObj.(*v1.Service)
+			newService := newObj.(*v1.Service)
+
+			log.Infof("Updated service '%s'", newService.GetName())
+			if oldService.Spec.Type == v1.ServiceTypeNodePort && newService.Spec.Type != v1.ServiceTypeNodePort {
+				log.Infof("Type of service '%s' changed ('%s' -> '%s'). Unexposing NodePorts...", newService.GetName(), oldService.Spec.Type, newService.Spec.Type)
+				for _, port := range oldService.Spec.Ports {
+					unexpose(port.NodePort)
+				}
+			} else if oldService.Spec.Type != v1.ServiceTypeNodePort && newService.Spec.Type == v1.ServiceTypeNodePort {
+				log.Infof("Type of service '%s' changed ('%s' -> '%s'). Exposing NodePorts...", newService.GetName(), oldService.Spec.Type, newService.Spec.Type)
+				for _, port := range newService.Spec.Ports {
+					expose(port.NodePort)
+				}
+			} else if oldService.Spec.Type == v1.ServiceTypeNodePort && newService.Spec.Type == v1.ServiceTypeNodePort {
+				log.Infof("Type of service '%s' didn't change ('%s' -> '%s'). Making sure that correct ports are exposed...", newService.GetName(), oldService.Spec.Type, newService.Spec.Type)
+				for _, port := range oldService.Spec.Ports {
+					unexpose(port.NodePort)
+				}
+
+				for _, port := range newService.Spec.Ports {
+					expose(port.NodePort)
+				}
+			} else {
+				log.Debugf("Neiter old nor new version of service '%s' was of type NodePort. Nothing to do...", newService.GetName())
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			service := obj.(*v1.Service)
+			log.Infof("Deleted service '%s' of type '%s'", service.GetName(), service.Spec.Type)
+			if service.Spec.Type != v1.ServiceTypeNodePort {
+				log.Debugf("Service '%s' is not of type NodePort. Nothing to do.", service.GetName())
+				return
+			}
+
+			for _, port := range service.Spec.Ports {
+				unexpose(port.NodePort)
+			}
+		},
+	})
+
+	return informer
+}
+
+type ExposeRequest struct {
+	Local  string
+	Remote string
+}
+
+type UnexposeRequest struct {
+	Local string
+}
+
+func expose(nodePort int32) {
+	exposeRequest := ExposeRequest{
+		Local:  fmt.Sprintf(":%d", nodePort),
+		Remote: fmt.Sprintf("%s:%d", vmVirtualIP, nodePort),
+	}
+	log.Infof("Exposing port (%s -> %s)", exposeRequest.Local, exposeRequest.Remote)
+
+	bin, err := json.Marshal(exposeRequest)
+	if err != nil {
+		log.Error(err)
```

Comment:
I'd do something similar to RoutesHandler and return errors from `expose` and `unexpose`, and log  them from the caller. You need at least a `return` here.